internal/painter/software/draw.go

```diff
@@ -34,29 +33,32 @@ func drawImage(c fyne.Canvas, img *canvas.Image, pos fyne.Position, frame fyne.S
 	bounds := img.Size()
 	width := internal.ScaleInt(c, bounds.Width)
 	height := internal.ScaleInt(c, bounds.Height)
-	tex := painter.PaintImage(img, c, width, height)
+	scaledX, scaledY := internal.ScaleInt(c, pos.X), internal.ScaleInt(c, pos.Y)
+
+	tmpImg := painter.PaintImage(img, c, width, height)
 
-	if img.FillMode == canvas.ImageFillStretch {
-		width = internal.ScaleInt(c, img.Size().Width)
-		height = internal.ScaleInt(c, img.Size().Height)
-		tex = resize.Resize(uint(width), uint(height), tex, resize.Lanczos3)
-	} else if img.FillMode == canvas.ImageFillContain {
+	if img.FillMode == canvas.ImageFillContain {
 		imgAspect := painter.GetAspect(img)
 		objAspect := float32(width) / float32(height)
 
 		if objAspect > imgAspect {
 			newWidth := int(float32(height) * imgAspect)
-			pos.X += (width - newWidth) / 2
+			scaledX += (width - newWidth) / 2
 			width = internal.ScaleInt(c, newWidth)
 		} else if objAspect < imgAspect {
 			newHeight := int(float32(width) / imgAspect)
-			pos.Y += (height - newHeight) / 2
+			scaledY += (height - newHeight) / 2
 			height = internal.ScaleInt(c, newHeight)
 		}
-		tex = resize.Resize(uint(width), uint(height), tex, resize.Lanczos3)
 	}
 
-	drawTex(c, pos, width, height, base, tex)
+	outBounds := image.Rect(scaledX, scaledY, scaledX+width, scaledY+height)
+	switch img.ScaleMode {
+	case canvas.ImageScaleSmooth:
```

Comment:
In line with the addition of panic lines in the GL driver I wonder if this should be in a default block? just a thought, open to discussion :)

Reply:
Same, I think default should panic, but is up to you (or the team ;-) ).

Reply:
I guess this one is still up for discussion? The behaviour is not consistent with the GL painter...

Reply:
default ?

Reply:
That would fix it I guess, maybe error and fallback like the Gl?